lib/std/os/linux/io_uring.zig

```diff
@@ -0,0 +1,877 @@
+    /// This can be used to chain causally dependent SQEs, and the chain can be arbitrarily long.
+    /// The tail of the chain is denoted by the first SQE that does not have this flag set.
+    /// This flag has no effect on previous SQEs, nor does it impact SQEs outside the chain.
+    /// This means that multiple chains can be executing in parallel, along with individual SQEs.
+    /// Only members inside the chain are serialized.
+    /// A chain will be broken if any SQE in the chain ends in error, where any unexpected result is
+    /// considered an error. For example, a short read will terminate the remainder of the chain.
+    pub fn link_with_next_sqe(self: *IO_Uring, sqe: *io_uring_sqe) void {
+        sqe.flags |= linux.IOSQE_IO_LINK;
+    }
+    
+    /// Like `link_with_next_sqe()` but stronger.
+    /// For when you don't want the chain to fail in the event of a completion result error.
+    /// For example, you may know that some commands will fail and may want the chain to continue.
+    /// Hard links are resilient to completion results, but are not resilient to submission errors.
+    pub fn hardlink_with_next_sqe(self: *IO_Uring, sqe: *io_uring_sqe) void {
+        sqe.flags |= linux.IOSQE_IO_HARDLINK;
+    }
+    
+    /// This creates a full pipeline barrier in the submission queue.
+    /// This SQE will not be started until previous SQEs complete.
+    /// Subsequent SQEs will not be started until this SQE completes.
+    /// In other words, this stalls the entire submission queue.
+    /// You should first consider using link_with_next_sqe() for more granular SQE sequence control.
+    pub fn drain_previous_sqes(self: *IO_Uring, sqe: *io_uring_sqe) void {
+        sqe.flags |= linux.IOSQE_IO_DRAIN;
+    }
+
+    /// Registers an array of file descriptors.
+    /// Every time a file descriptor is put in an SQE and submitted to the kernel, the kernel must
+    /// retrieve a reference to the file, and once I/O has completed the file reference must be
+    /// dropped. The atomic nature of this file reference can be a slowdown for high IOPS workloads.
+    /// This slowdown can be avoided by pre-registering file descriptors.
+    /// To refer to a registered file descriptor, IOSQE_FIXED_FILE must be set in the SQE's flags,
+    /// and the SQE's fd must be set to the index of the file descriptor in the registered array.
+    /// Registering file descriptors will wait for the ring to idle.
+    /// Files are automatically unregistered by the kernel when the ring is torn down.
+    /// An application need unregister only if it wants to register a new array of file descriptors.
+    pub fn register_files(self: *IO_Uring, fds: []const os.fd_t) !void {
+        assert(self.fd >= 0);
+        comptime assert(@sizeOf(os.fd_t) == @sizeOf(c_int));
+        const res = linux.io_uring_register(
+            self.fd,
+            .REGISTER_FILES,
+            @ptrCast(*const c_void, fds.ptr),
+            @truncate(u32, fds.len)
```

Comment:
I don't think `@truncate` is valid here: unlike for reads/writes there is no partial success here

Reply:
I don't understand? `fds.len` is simply a `usize` and we need to truncate that down to `u32`.

Reply:
We're not truncating a syscall `res` here.

Reply:
the question is what happens if `fds.len` is e.g. `1<<34`: with `@truncate` we cut it off and pass `0` instead; with `@intCast` we would get safety-checked behaviour in safe release modes or undefined behaviour in release-fast.
I think for this case we want `@intCast` and to document that it shouldn't be larger than a `u32`.

Reply:
Thanks, agreed. It's now an `@intCast` as well as all other `@truncate` calls. By the way, the std lib also uses @truncate at present in many of these places.

Do we need to document that `fds.len` can't be more than 4 billion files? Would any kernel even allow a user to get that far and open so many files?